src/filetree.rs

```diff
@@ -302,40 +353,82 @@ pub(crate) fn apply_diff(
     let opts = opts.unwrap_or(&default_opts);
     cleanup_tmp(destdir).context("cleaning up temporary files")?;
 
-    // Write new and changed files
-    for pathstr in diff.additions.iter().chain(diff.changes.iter()) {
+    // Get updates based on dir
+    let mut updates = std::collections::HashSet::new();
+    for pathstr in diff
+        .additions
+        .iter()
+        .chain(diff.changes.iter())
+        .chain(diff.removals.iter())
+    {
         let path = Path::new(pathstr);
         if let Some(parent) = path.parent() {
-            destdir.ensure_dir_all(parent, DEFAULT_FILE_MODE)?;
+            updates.insert(parent.to_string_lossy().into_owned());
         }
-        let destp = tmpname_for_path(path);
-        srcdir
-            .copy_file_at(path, destdir, destp.as_path())
-            .with_context(|| format!("writing {}", &pathstr))?;
     }
-    // Ensure all of the new files are written persistently to disk
-    if !opts.skip_sync {
-        syncfs(destdir)?;
-    }
-    // Now move them all into place (TODO track interruption)
-    for path in diff.additions.iter().chain(diff.changes.iter()) {
-        let pathtmp = tmpname_for_path(path);
+    let efipath = destdir.recover_path()?;
+    for path in &updates {
+        let new_diff = diff_with_prefix(diff, path)?;
```

Comment:
@cgwalters one thing want to confirm is that, this logic is using apply/exchange based on updates dir's parent, but according to @travier 's suggestion, should copy first dir as temp (like `BOOT/foo` -> `BOOT.tmp/foo`) and remember, then do remove/addition/change files in temp dir, finally scan temp and do exchange. WDYT?